grpc/authz/authz.go

```diff
@@ -29,6 +30,87 @@ type Base struct {
 	ValidateScopeFunc         func(tokenClaims []byte, scopes []string) error
 	Audience                  string
 	DisableAudienceValidation bool
+	// HSKey for use with HS algorithms
+	HSKey []byte
+	// RSAPublicKey for use with RS algorithms
+	RSAPublicKey *rsa.PublicKey
+}
+
+// ConfigOption for setting optional values
+type ConfigOption func(*Config)
+
+// WithScopeMapping sets the ScopeMapping option
+func WithScopeMapping(scopeMap map[string][]string) ConfigOption {
+	return func(args *Config) { args.ScopeMapping = scopeMap }
+}
+
+// WithValidateScopeFunc sets the ValidateScopeFunc option
+func WithValidateScopeFunc(scopeFunc func(tokenClaims []byte, scopes []string) error) ConfigOption {
+	return func(args *Config) { args.ValidateScopeFunc = scopeFunc }
+}
+
+// WithAudience sets the audience
+func WithAudience(aud string) ConfigOption {
+	return func(args *Config) { args.Audience = aud }
+}
+
+// WithDisableAudienceValidation sets the WithDisableAudienceValidation option
+func WithDisableAudienceValidation(disable bool) ConfigOption {
+	return func(args *Config) { args.DisableAudienceValidation = disable }
+}
+
+// WithHSKey sets the HS key
+func WithHSKey(hKey []byte) ConfigOption {
+	return func(args *Config) { args.HSKey = hKey }
+}
+
+// WithRSAPubKey sets the RSA public key
+func WithRSAPubKey(rsaPubKey *rsa.PublicKey) ConfigOption {
+	return func(args *Config) { args.RSAPublicKey = rsaPubKey }
+}
+
+// NewConfig returns a new config with options
+func NewConfig(algo jwt.Algorithm, opts ...ConfigOption) *Config {
+	defaultConfig := &Config{
+		Algorithm:         algo,
+		ValidateScopeFunc: func(tokenClaims []byte, scopes []string) error { return nil },
+	}
+	for _, opt := range opts {
+		opt(defaultConfig)
+	}
+	return defaultConfig
+}
+
+// AuthFunc authorization function
+func (c *Config) AuthFunc(ctx context.Context) (context.Context, error) {
+	token, protected, err := c.doProtected(ctx)
```

Comment:
I'm a little concerned about hitting any of the jwt footguns. I'd really love some documentation on steps this package takes to make sure we don't shoot ourselves in the foot with jwt, especially to avoid having to dig through to go-jwt docs to find some of this out.

Questions I have (but not exhaustive):

Do we allow only safe algos?
Do we block null/none?

What does protected even mean here?

Reply:
_Do we allow only safe algos?_
as far as I can tell, yes. https://github.com/cristalhq/jwt/blob/f352884e5ec5933cccb57299e52adf59c907da3b/parse.go#L69

_Do we block null/none?_
as far as I can tell, yes. https://github.com/cristalhq/jwt/blob/f352884e5ec5933cccb57299e52adf59c907da3b/parse.go#L69

_What does protected even mean here?_
ah yes, good question/catch. `doProtected` needs to be updated. I will do that. `doProtected` is intended to check if the gRPC method should be protected by auth or not.